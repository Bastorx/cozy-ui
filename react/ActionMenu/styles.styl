@require '../../stylus/components/action-menu'

.c-actionmenu
    @extends $actionmenu

.c-actionmenu-header
    @extends $actionmenu-header

.c-actionmenu-item
    @extends $actionmenu-item
