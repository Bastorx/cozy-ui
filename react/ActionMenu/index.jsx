import React, { useCallback, useContext } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import ClickAwayListener from '@material-ui/core/ClickAwayListener'
import styles from './styles.styl'
import { Media, Bd, Img } from '../Media'
import BottomDrawer from '../BottomDrawer'
import withBreakpoints from '../helpers/withBreakpoints'
import Popper from '@material-ui/core/Popper'
import { getCssVariableValue } from '../utils/color'
import PopperContainerContext from '../PopperContainerContext'
import Radio from '../Radio'
import { spacingProp } from '../Stack'

const ActionMenuWrapper = ({
  inline,
  onClose,
  anchorElRef,
  containerElRef,
  placement,
  preventOverflow,
  children
}) => {
  const getElementFromRefCallback = ref =>
    useCallback(() => {
      return ref ? ref.current : undefined
    }, [ref])

  const popperContainerRef = useContext(PopperContainerContext)
  const getAnchorElement = getElementFromRefCallback(anchorElRef)

  const containerRef = popperContainerRef || containerElRef
  const getContainerElement = getElementFromRefCallback(containerRef)
  const normalOverflowModifiers = {
    preventOverflow: { enabled: false },
    hide: { enabled: false }
  }

  return inline ? (
    <Popper
      anchorEl={getAnchorElement}
      container={getContainerElement}
      modifiers={preventOverflow ? null : normalOverflowModifiers}
      open
      placement={placement}
      style={{
        zIndex: getCssVariableValue('zIndex-popover')
      }}
    >
      <ClickAwayListener onClickAway={onClose}>{children}</ClickAwayListener>
    </Popper>
  ) : (
    <BottomDrawer onClose={onClose}>{children}</BottomDrawer>
  )
}

const ActionMenu = ({
  children,
  autoclose,
  className,
  onClose,
  placement,
  preventOverflow,
  anchorElRef,
  containerElRef,
  breakpoints: { isDesktop }
}) => {
  const shouldDisplayInline = isDesktop
  const containerRef = React.createRef()
  return (
    <div
      className={className}
      ref={containerRef}
      onClick={autoclose ? onClose : null}
    >
      <ActionMenuWrapper
        onClose={onClose}
        inline={shouldDisplayInline}
        anchorElRef={anchorElRef || containerRef}
        containerElRef={containerElRef}
        placement={placement}
        preventOverflow={preventOverflow}
      >
        <div
          className={cx(styles['c-actionmenu'], {
            [styles['c-actionmenu--inline']]: shouldDisplayInline
          })}
        >
          {React.Children.map(children, child =>
            child && child.type === ActionMenuHeader && isDesktop ? null : child
          )}
        </div>
      </ActionMenuWrapper>
    </div>
  )
}

ActionMenu.propTypes = {
  /** Extra class */
  className: PropTypes.string,
  /** What to do on close */
  onClose: PropTypes.func,
  /** Whether the menu should automatically close itself when an item is clicked */
  autoclose: PropTypes.bool,
  /** Controls the placement of the menu on desktop */
  placement: PropTypes.oneOf([
    'bottom-end',
    'bottom-start',
    'bottom',
    'left-end',
    'left-start',
    'left',
    'right-end',
    'right-start',
    'right',
    'top-end',
    'top-start',
    'top'
  ]),
  /** Will keep the menu visible when scrolling */
  preventOverflow: PropTypes.bool,
  /** The reference element for the menu placement and overflow prevention. */
  anchorElRef: PropTypes.object,
  /** ActionMenu will be rendered inside the elemnt of this ref. Useful when rendering inside Modals for example. */
  containerElRef: PropTypes.object
}

ActionMenu.defaultProps = {
  placement: 'bottom-start',
  preventOverflow: false,
  autoclose: false
}

const ActionMenuHeader = ({ children, className }) => {
  return (
    <div className={cx(styles['c-actionmenu-header'], className)}>
      {children}
    </div>
  )
}

ActionMenuHeader.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string
}

const ActionMenuItem = ({
  left,
  children,
  right,
  onClick,
  className,
  contentSpacing
}) => {
  return (
    <Media
      className={cx(styles['c-actionmenu-item'], className)}
      onClick={onClick}
      align="top"
    >
      {left && <Img className="u-mh-1">{left}</Img>}
      <Bd
        className={cx(left ? 'u-mr-1' : 'u-mh-1', `u-stack-${contentSpacing}`)}
      >
        {children}
      </Bd>
      {right && <Img className="u-mr-1">{right}</Img>}
    </Media>
  )
}

const ActionMenuRadio = props => {
  return <Radio {...props} className={styles['c-actionmenu-radio']} />
}

ActionMenuItem.propTypes = {
  left: PropTypes.node,
  right: PropTypes.node,
  children: PropTypes.node,
  onClick: PropTypes.func,
  className: PropTypes.string,
  /** Controls spacing between between children of the ActionMenuItem */
  contentSpacing: spacingProp
}

ActionMenuItem.defaultProps = {
  contentSpacing: 'xs'
}

export default withBreakpoints()(ActionMenu)
export { ActionMenuHeader, ActionMenuItem, ActionMenuRadio }
