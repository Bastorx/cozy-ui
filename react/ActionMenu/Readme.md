Use an ActionMenu to show a list of actions. ActionMenus automatically switch their display between desktop and mobile.

### Classic

```
import ActionMenu, { ActionMenuItem, ActionMenuRadio } from 'cozy-ui/transpiled/react/ActionMenu';
import DropdownButton from 'cozy-ui/transpiled/react/DropdownButton';
import Icon from 'cozy-ui/transpiled/react/Icon';
import { Text, Caption } from 'cozy-ui/transpiled/react/Text';

initialState = { menuDisplayed: isTesting() };

const showMenu = () => setState({ menuDisplayed: true });
const hideMenu = () => setState({ menuDisplayed: false });

<div>
  <DropdownButton onClick={showMenu}>Show action menu</DropdownButton>
  {state.menuDisplayed &&
    <ActionMenu
      onClose={hideMenu}>
      <ActionMenuItem left={<Icon icon='file' />} right={<Icon icon='warning' />}>Item 1</ActionMenuItem>
      <ActionMenuItem left={<ActionMenuRadio />}>Item 2</ActionMenuItem>
      <ActionMenuItem left={<Icon icon='file' />}>
        <Text>
          Item 3
        </Text>
        <Caption>
          Descriptive text to elaborate on what item 3 does.
        </Caption>
      </ActionMenuItem>
  </ActionMenu>}
</div>
```

### With Header (mobile only)

A header can be used to provide context on the menu actions. Since on
desktop, we display a popper and not a BottomDrawer, context for the
user is not lost, so the ActionMenuHeader would be redundant. This is
why it is not rendered unless we are on mobile.

```
import ActionMenu, { ActionMenuItem, ActionMenuHeader } from './index';
import DropdownButton from 'cozy-ui/transpiled/react/DropdownButton';
import Icon from '../Icon';
import Filename from '../Filename';

initialState = { menuDisplayed: isTesting() };

const showMenu = () => setState({ menuDisplayed: true });
const hideMenu = () => setState({ menuDisplayed: false });

<div>
  <DropdownButton onClick={showMenu}>Show action menu</DropdownButton>
  {state.menuDisplayed &&
    <ActionMenu
      onClose={hideMenu}>
      <ActionMenuHeader>
        <Filename icon="file" filename="my_awesome_paper" extension=".pdf" />
      </ActionMenuHeader>
      <ActionMenuItem left={<Icon icon='file' />}>Item 1</ActionMenuItem>
      <ActionMenuItem left={<Icon icon='right' />}>Item 2</ActionMenuItem>
      <ActionMenuItem left={<Icon icon='file' />}>Item 3</ActionMenuItem>
  </ActionMenu>}
</div>
```

### With dedicated click handler

```
import ActionMenu, { ActionMenuItem, ActionMenuHeader } from './index';
import DropdownButton from 'cozy-ui/transpiled/react/DropdownButton';
import Icon from '../Icon';
import Filename from '../Filename';

initialState = { menuDisplayed: isTesting() };

const showMenu = () => setState({ menuDisplayed: true });
const hideMenu = () => setState({ menuDisplayed: false });

<div>
  <DropdownButton onClick={showMenu}>Show action menu</DropdownButton>
  {state.menuDisplayed &&
    <ActionMenu
      onClose={hideMenu}>
      <ActionMenuHeader>
        <Filename icon="file" filename="my_awesome_paper" extension=".pdf" />
      </ActionMenuHeader>
      <ActionMenuItem onClick={() => alert('click')}left={<Icon icon='file' />}>Item 1</ActionMenuItem>
  </ActionMenu>}
</div>
```

### Placement

The `placement` and `anchorElRef` prop can be used to control the placement of the menu on desktop. `anchorElRef` should be a ref to a DOM element and not a react component.

```
import ActionMenu, { ActionMenuItem } from 'cozy-ui/transpiled/react/ActionMenu';
import DropdownButton from 'cozy-ui/transpiled/react/DropdownButton';
import Icon from 'cozy-ui/transpiled/react/Icon';

const testRef = React.createRef();

initialState = { menuDisplayed: isTesting() };

const showMenu = () => setState({ menuDisplayed: true });
const hideMenu = () => setState({ menuDisplayed: false });

const anchorRef = React.createRef();

<div>
  <DropdownButton onClick={showMenu} ref={anchorRef}>Show action menu</DropdownButton>
  {state.menuDisplayed &&
    <ActionMenu
      anchorElRef={anchorRef}
      placement="bottom-end"
      buttonRef={testRef}
      onClose={hideMenu}>
      <ActionMenuItem left={<Icon icon='file' />}>Item 1</ActionMenuItem>
  </ActionMenu>}
</div>
```

### ActionMenu inside custom fixed elements

The ActionMenu component is rendered at the root of the DOM tree using a [Portal](https://reactjs.org/docs/portals.html) and a fixed z-index. Since Modals for example have a higher z-index than ActionMenus, an ActionMenu _inside_ a Modal would be displayed behind it, instead of over it. To fix this problem, the Popper instance
is given a reference to the Modal DOM node for its `container` option via a special
React context.

Any z-indexed element that will display action menus should use this technique to
provide a ref to a node inside this stacking context (CSS context, not React
context).

```
import { useRef, useState, useCallback } from 'react'
import PopperContainerContext from '../PopperContainerContext'
import ActionMenu, { ActionMenuItem, ActionMenuHeader } from './index';
import Icon from '../Icon';

const MyActionMenu = ({ anchorRef, buttonRef, onClose }) => {
  return <ActionMenu
      anchorElRef={anchorRef}
      placement="bottom-end"
      buttonRef={buttonRef}
      onClose={onClose}>
      <ActionMenuItem onClick={onClose} left={<Icon icon='file' />}>Item 1</ActionMenuItem>
  </ActionMenu>
}

const MyFixedElement = () => {
  const ref = useRef()
  const [shown, setShown] = useState(isTesting())
  const buttonRef = useRef()
  const show = useCallback(() => setShown(true), [setShown])
  const hide = useCallback(() => setShown(false), [setShown])
  return <PopperContainerContext.Provider value={ref}>
    <div style={{ position: 'fixed', height: 100, width: 100, bottom: 0, top: 0, margin: 'auto', left: 0, right: 0, zIndex: 1 }}>
      <div ref={ref} />
      <button onClick={show} ref={buttonRef}>Open action menu</button>
      {shown ? <MyActionMenu onClose={hide} buttonRef={buttonRef}/> : null}
    </div>
  </PopperContainerContext.Provider>
};

initialState = { shown: isTesting() };

<>
  <button onClick={() => setState({ shown: !state.shown })}>
    Toggle custom fixed element
  </button>
  { state.shown && <MyFixedElement /> }
</>
```

### ActionMenu within modals

The above technique is used in Modals, nothing is required from the developer
for ActionMenus inside modals to work correctly.

```
import { useState, useRef } from 'react'
import ActionMenu, { ActionMenuItem, ActionMenuHeader } from './index';
import Icon from '../Icon';
import Modal, { ModalDescription } from '../Modal';
import Filename from '../Filename';

initialState = { modalDisplayed: isTesting() };

const showModal = () => setState({ modalDisplayed: true });
const hideModal = () => setState({ modalDisplayed: false });


const ExampleModalContent = () => {
  const [menuDisplayed, setMenuDisplayed] = useState(isTesting())
    const showMenu = () => setMenuDisplayed(true);
    const hideMenu = () => setMenuDisplayed(false);
    return (
      <ModalDescription>
          <button onClick={showMenu}>Show action menu</button>
        {menuDisplayed &&
          <ActionMenu
            onClose={hideMenu}>
            <ActionMenuHeader>
              <Filename icon="file" filename="my_awesome_paper" extension=".pdf" />
            </ActionMenuHeader>
            <ActionMenuItem onClick={() => alert('click')}left={<Icon icon='file' />}>Item 1</ActionMenuItem>
        </ActionMenu>}
      </ModalDescription>
    )
}

<div>
  <button onClick={showModal}>Show modal</button>
  {state.modalDisplayed &&
    <Modal dismissAction={hideModal}>
      <ExampleModalContent />
    </Modal>
  }
</div>
```

### preventOverflow

Set `preventOverflow` to `true` to keep the ActionMenu visible on desktop, even if `anchorElRef` is outside the viewport.
