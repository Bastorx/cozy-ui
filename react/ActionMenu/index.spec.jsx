import React from 'react'
import { mount } from 'enzyme'

import ActionMenu, { ActionMenuItem } from './'
import { fixPopperTesting } from '../Popper/testing'

describe('ActionMenu', () => {
  fixPopperTesting()

  it('should support null children', () => {
    const comp = mount(
      <ActionMenu onClose={jest.fn()}>
        <ActionMenuItem>Item 1</ActionMenuItem>
        {null}
      </ActionMenu>
    )
    expect(
      comp
        .find(ActionMenuItem)
        .parent()
        .getElement()
    ).toMatchSnapshot()
  })

  it('should support auto-closing the menu', () => {
    const closeMenu = jest.fn()
    const menuAction1 = jest.fn()
    const menuAction2 = jest.fn()
    const menuActionStoppingPropagation = e => {
      e.stopPropagation()
      menuAction2()
    }

    const comp = mount(
      <ActionMenu onClose={closeMenu} autoclose>
        <ActionMenuItem onClick={menuAction1}>Item 1</ActionMenuItem>
        <ActionMenuItem onClick={menuActionStoppingPropagation}>
          Item 2
        </ActionMenuItem>
      </ActionMenu>
    )

    comp
      .find(ActionMenuItem)
      .at(1)
      .simulate('click')
    expect(menuAction2).toHaveBeenCalled()
    expect(closeMenu).not.toHaveBeenCalled()
    comp
      .find(ActionMenuItem)
      .at(0)
      .simulate('click')
    expect(menuAction1).toHaveBeenCalled()
    expect(closeMenu).toHaveBeenCalled()
  })
})
