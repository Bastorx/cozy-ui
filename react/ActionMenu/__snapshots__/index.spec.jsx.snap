// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionMenu should support null children 1`] = `
<div
  className="styles__c-actionmenu___22Fp1 styles__c-actionmenu--inline___1SXZa"
>
  <ActionMenuItem
    contentSpacing="xs"
  >
    Item 1
  </ActionMenuItem>
</div>
`;
